src/opencl_fvde_fmt_plug.c

```diff
@@ -0,0 +1,350 @@
+ * Copyright (c) 2013 Lukas Odzioba <ukasz at openwall dot net> and it is
+ * hereby released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ */
+#ifdef HAVE_OPENCL
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_opencl_fvde;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_opencl_fvde);
+#else
+
+#include <string.h>
+
+#include "misc.h"
+#include "arch.h"
+#include "fvde_common.h"
+#include "common.h"
+#include "stdint.h"
+#include "formats.h"
+#include "options.h"
+#include "common-opencl.h"
+
+#define FORMAT_LABEL		"FVDE-OpenCL"
+#define ALGORITHM_NAME		"PBKDF2-SHA256 OpenCL AES"
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1001
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            MEM_ALIGN_WORD
+#define SALT_SIZE		sizeof(*cur_salt)
+#define SALT_ALIGN		1
+#define PLAINTEXT_LENGTH	55
+#define KERNEL_NAME		"pbkdf2_sha256_kernel"
+#define SPLIT_KERNEL_NAME	"pbkdf2_sha256_loop"
+
+#define HASH_LOOPS		(13*71) // factors 13, 13, 71
+#define ITERATIONS		12000
+
+typedef struct {
+	uint8_t length;
+	uint8_t v[PLAINTEXT_LENGTH];
+} pass_t;
+
+typedef struct {
+	uint32_t hash[8];
+} crack_t;
+
+typedef struct {
+	uint8_t length;
+	uint8_t salt[115];
+	uint32_t rounds;
+} salt_t;
+
+typedef struct {
+	uint32_t ipad[8];
+	uint32_t opad[8];
+	uint32_t hash[8];
+	uint32_t W[8];
+	uint32_t rounds;
+} state_t;
+
+static pass_t *host_pass;			      /** plain ciphertexts **/
+static salt_t *host_salt;			      /** salt **/
+static crack_t *host_crack;			      /** hash**/
+static cl_int cl_error;
+static cl_mem mem_in, mem_out, mem_salt, mem_state;
+static cl_kernel split_kernel;
+static struct fmt_main *self;
+
+static int *cracked, cracked_size;
+static int any_cracked;
+static fvde_custom_salt *cur_salt;
+
+#define STEP			0
+#define SEED			1024
+
+static const char * warn[] = {
+        "xfer: ",  ", init: " , ", crypt: ", ", res xfer: "
+};
+
+static int split_events[] = { 2, -1, -1 };
+
+// This file contains auto-tuning routine(s). Has to be included after formats definitions.
+#include "opencl-autotune.h"
+#include "memdbg.h"
+
+static void create_clobj(size_t kpc, struct fmt_main *self)
+{
+#define CL_RO CL_MEM_READ_ONLY
+#define CL_WO CL_MEM_WRITE_ONLY
+#define CL_RW CL_MEM_READ_WRITE
+
+#define CLCREATEBUFFER(_flags, _size, _string)\
+	clCreateBuffer(context[gpu_id], _flags, _size, NULL, &cl_error);\
+	HANDLE_CLERROR(cl_error, _string);
+
+#define CLKERNELARG(kernel, id, arg, msg)\
+	HANDLE_CLERROR(clSetKernelArg(kernel, id, sizeof(arg), &arg), msg);
+
+	host_pass = mem_calloc(kpc, sizeof(pass_t));
+	host_crack = mem_calloc(kpc, sizeof(crack_t));
+	host_salt = mem_calloc(1, sizeof(salt_t));
+	cracked = mem_calloc(kpc, sizeof(*cracked));
+	cracked_size = kpc; // XXX is this correct?
```

Comment:
Actually I think `sizeof(int)` is easier to "follow". But it's obviously the same thing.